File: src/test/programmers/java/level2/TargetNumber.java
package test.programmers.java.level2;

public class TargetNumber {

    static int length;
    static int answer = 0;

    public int solution(int[] numbers, int target) {
        length = numbers.length;

        dfs(0, 0, numbers, target);

        return answer;
    }

    static void dfs(int idx, int result, int[] numbers, int target) {
        if (idx == length) {
            if (result == target) {
                answer++;
            }

            return;
        }

        dfs(idx + 1, result + numbers[idx], numbers, target);
        dfs(idx + 1, result - numbers[idx], numbers, target);
    }
}

File: src/test/programmers/python/level2/TargetNumber.py
def solution(numbers, target):
    length = len(numbers)
    answer = 0

    def dfs(idx, result):
        if idx == length:
            if target == result:
                nonlocal answer
                answer += 1
            return
        else:
            dfs(idx + 1, result + numbers[idx])
            dfs(idx + 1, result - numbers[idx])

    dfs(0, 0)

    return answer

print(solution([1, 1, 1, 1, 1], 3))
